src/tasks/publish/index.ts

```diff
@@ -0,0 +1,226 @@
+  const selectedNetwork = bre.network.name
+  // ### Todo: Add logic to guess app name from file?
+  const appName =
+    typeof aragonConfig.appName === 'string'
+      ? aragonConfig.appName
+      : typeof aragonConfig.appName === 'object'
+      ? aragonConfig.appName[selectedNetwork]
+      : undefined
+  if (!appName)
+    throw new BuidlerPluginError(`appName must be defined in buidler.config`)
+  const contractName = getMainContractName()
+
+  // Initialize clients
+  const networkConfig = bre.network.config as HttpNetworkConfig
+  const provider = new ethers.providers.Web3Provider(
+    bre.web3.currentProvider,
+    networkConfig.ensAddress && {
+      name: bre.network.name,
+      chainId: networkConfig.chainId || 5555,
+      ensAddress: networkConfig.ensAddress
+    }
+  )
+
+  const prevVersion = await _getLastestVersionIfExists(appName, provider)
+
+  const { bump, nextVersion } = parseAndValidateBumpOrVersion(
+    bumpOrVersion,
+    prevVersion ? prevVersion.version : undefined
+  )
+  logMain(`Applying version bump ${bump}, next version: ${nextVersion}`)
+
+  // Using let + if {} block instead of a ternary operator
+  // to assign value and log status to console
+  let contractAddress: string
+  if (onlyContent) {
+    logMain('No contract used for this version')
+    contractAddress = zeroAddress
+  } else if (existingContractAddress) {
+    logMain('Using provided contract address')
+    contractAddress = existingContractAddress
+  } else if (!prevVersion || bump === 'major') {
+    logMain('Deploying new implementation contract')
+    contractAddress = await _deployMainContract(contractName, noVerify, bre)
+  } else {
+    logMain('Reusing contract from previous version')
+    contractAddress = prevVersion.contractAddress
+  }
+  logMain(`contractAddress: ${contractAddress}`)
+
+  // Prepare release directory
+  // npm run build must create: index.html, src.js, script.js
+  logMain(`Building app front-end at ${appSrcPath}`)
+  await execa('npm', ['run', 'build'], { cwd: appSrcPath })
+
+  // Generate and validate Aragon artifacts, release files
+  logMain(`Generating Aragon app artifacts`)
+  await generateArtifacts(distPath, bre)
+  validateArtifacts(distPath)
+
+  // Upload release directory to IPFS
+  logMain('Uploading release assets to IPFS...')
+  const contentHash = await uploadReleaseToIpfs(distPath, {
+    ipfsProvider,
+    ignorePatterns: ipfsIgnore
+  })
+  logMain(`Release assets uploaded to IPFS: ${contentHash}`)
+
+  // Generate tx to publish new app to aragonPM
+  const versionInfo = {
+    version: nextVersion,
+    contractAddress,
+    contentUri: apm.toContentUri('ipfs', contentHash)
+  }
+
+  const txData = await apm.publishVersion(appName, versionInfo, provider, {
+    managerAddress
+  })
+
+  logMain(
+    `Successfully generated TX data for publishing ${appName} version ${nextVersion}
+
+to: ${txData.to}
+data: ${apm.encodePublishVersionTxData(txData)}
+`
+  )
```

Comment:
Since we have the provider and we're already deploying a contract, it seems weird we wouldn't send the publish tx here as well (with an option to configure it to be a dry run).

Reply:
Totally agree, in a small effort to keep the changes minimal I left this to the next PR

Reply:
Let's do that; could you track it in #77?

Reply:
Added